packages/util-dynamodb/src/marshall.ts

```diff
@@ -36,19 +37,51 @@ export interface marshallOptions {
+  options?: marshallOptions
+): AttributeValue.NSMember;
 export function marshall(data: Set<string>, options?: marshallOptions): AttributeValue.SSMember;
-export function marshall(data: Set<number>, options?: marshallOptions): AttributeValue.NSMember;
 export function marshall(data: Set<NativeAttributeBinary>, options?: marshallOptions): AttributeValue.BSMember;
-export function marshall<M extends { [K in keyof M]: NativeAttributeValue }>(
-  data: M,
-  options?: marshallOptions
-): Record<string, AttributeValue>;
-export function marshall<L extends NativeAttributeValue[]>(data: L, options?: marshallOptions): AttributeValue[];
-export function marshall(data: string, options?: marshallOptions): AttributeValue.SMember;
-export function marshall(data: number, options?: marshallOptions): AttributeValue.NMember;
 export function marshall(data: NativeAttributeBinary, options?: marshallOptions): AttributeValue.BMember;
-export function marshall(data: null, options?: marshallOptions): AttributeValue.NULLMember;
 export function marshall(data: boolean, options?: marshallOptions): AttributeValue.BOOLMember;
+export function marshall(data: number | NumberValue | bigint, options?: marshallOptions): AttributeValue.NMember;
+export function marshall(data: string, options?: marshallOptions): AttributeValue.SMember;
+export function marshall(data: boolean, options?: marshallOptions): AttributeValue.BOOLMember;
+export function marshall<O extends { convertTopLevelContainer: true }>(
+  data: NativeAttributeValue[],
+  options: marshallOptions & O
+): AttributeValue.LMember;
+export function marshall<O extends { convertTopLevelContainer: false }>(
+  data: NativeAttributeValue[],
+  options: marshallOptions & O
+): AttributeValue[];
+export function marshall<O extends { convertTopLevelContainer: boolean }>(
+  data: NativeAttributeValue[],
+  options: marshallOptions & O
+): AttributeValue[] | AttributeValue.LMember;
+export function marshall(data: NativeAttributeValue[], options?: marshallOptions): AttributeValue[];
+export function marshall<O extends { convertTopLevelContainer: true }>(
+  data: Map<string, NativeAttributeValue> | Record<string, NativeAttributeValue>,
+  options: marshallOptions & O
+): AttributeValue.MMember;
+export function marshall<O extends { convertTopLevelContainer: false }>(
+  data: Map<string, NativeAttributeValue> | Record<string, NativeAttributeValue>,
+  options: marshallOptions & O
+): Record<string, AttributeValue>;
+export function marshall<O extends { convertTopLevelContainer: boolean }>(
+  data: Map<string, NativeAttributeValue> | Record<string, NativeAttributeValue>,
+  options: marshallOptions & O
+): Record<string, AttributeValue> | AttributeValue.MMember;
+export function marshall(
+  data: Map<string, NativeAttributeValue> | Record<string, NativeAttributeValue>,
+  options?: marshallOptions
+): Record<string, AttributeValue>;
+export function marshall(data: any, options?: marshallOptions): any;
```

Comment:
added this `any => any` signature to prevent inputs typed as `any` from matching to some other signature.

Reply:
https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAQwBQDcBciwgLYBGApgE4CU2A7ANwBQoksCKG2uhpA2gLoWIBsdBtHhI0WFGACePPgFY6AekWJhTMa0lS+AFiHgRzcdnABrMHADuYPsmmIA3rUQvEJIlBAkkABjoBfWloIBABnKER0AEZsOUQAXhZOAHIfZN5qRGVEPGQoCAALIlDEUJgAczA8ryItHmCwiPQAJioElgd-FBK7bUzsyxgoArgQCOHa1F6yRBgwAAcx0oqqz3cAGkRhmBLc-KKSqDgtotUYEnDlyur3NyJy5BIAEwAbYpK4YERLArzZiJ2ADogmpRIgCJp2MRyFR9IwwRCJFCuLxsIJ6AZ1ODNL1ZLE4YYkIjsNNsHoMfDmMTEGYLNZbPYnK47mtfAEgiEwJcAI4xATtCGcHwZBpciLc1qIOKJQWpdJkOicnkAZjJAtQnW6WgVWRUIRI7mgLykOTyhWKW2OZWua1qvUQAFoAHyIHSzMVEZBPRCfE61YDnS7W1Y1QFAA